api/src/permissions/utils/get-permissions-for-share.ts

```diff
@@ -0,0 +1,275 @@
+import { schemaPermissions } from '@directus/system-data';
+import type { Accountability, Filter, Permission, SchemaOverview } from '@directus/types';
+import { set, uniq } from 'lodash-es';
+import { fetchAllowedFieldMap } from '../modules/fetch-allowed-field-map/fetch-allowed-field-map.js';
+import type { Context } from '../types.js';
+import { fetchShareInfo } from './fetch-share-info.js';
+import { mergePermissions } from './merge-permissions.js';
+import { fetchPermissions } from '../lib/fetch-permissions.js';
+import { fetchPolicies } from '../lib/fetch-policies.js';
+import { fetchRolesTree } from '../lib/fetch-roles-tree.js';
+import { reduceSchema } from '../../utils/reduce-schema.js';
+import { fetchGlobalAccess } from '../modules/fetch-global-access/fetch-global-access.js';
+
+export async function getPermissionsForShare(
+	accountability: Pick<Accountability, 'share' | 'ip'>,
+	collections: string[] | undefined,
+	context: Context,
+): Promise<Permission[]> {
+	const defaults: Permission = {
+		action: 'read',
+		collection: '',
+		permissions: {},
+		policy: null,
+		validation: null,
+		presets: null,
+		fields: null,
+	};
+
+	const { collection, item, role, user_created } = await fetchShareInfo(accountability.share!, context);
+
+	const userAccountability: Accountability = {
+		user: user_created.id,
+		role: user_created.role,
+		roles: await fetchRolesTree(user_created.role, context.knex),
+		admin: false,
+		app: false,
+		ip: accountability.ip,
+	};
+
+	// Fallback to public accountability so merging later on has no issues
+	const shareAccountability: Accountability = {
+		user: null,
+		role: role,
+		roles: await fetchRolesTree(role, context.knex),
+		admin: false,
+		app: false,
+		ip: accountability.ip,
+	};
+
+	const { admin: shareIsAdmin } = await fetchGlobalAccess(shareAccountability, context.knex);
```

Comment:
We can potentially speed this up by requesting `fetchGlobalAccess`,`getPermissionsForAccountability` and `fetchAllowedFieldMap` in parallel via `Promise.all` as they are independent queries.